### PR TITLE
[DS-134]: Button loading state

### PR DESCRIPTION
**Pull request by michalisKout**
## Description

- adds button loading state handling for async operators
- introduces new spinner loader 
- refactors Loader component to make it more scalable for the upcoming loaders

[jira task](https://orfium.atlassian.net/secure/RapidBoard.jspa?rapidView=40&projectKey=DS&modal=detail&selectedIssue=DS-134)

## Screenshot

ezgif com-gif-maker (2)

Async Buttons => https://ictinus-fix-ds-134-o0590dditrk.herokuapp.com/?path=/story/design-system-button--async-button
Refactored Loaders => https://ictinus-fix-ds-134-o0590dditrk.herokuapp.com/?path=/story/design-system-loaders--available-loaders

## Test Plan

- updates snapshots

### CODE REVIEW

**Comment by michalisKout** [comment]
@panvourtsis 
Should we create a new task for the new spinner loader or should I create it in this PR? 
Also should we include the async handling for IconButton too?

**Comment by panvourtsis** [comment]
i am ok for handling it here

**Comment by michalisKout** [comment]
> i am ok for handling it here

ok! Converting the PR to draft to implement the spinner

File: src/components/Button/Button.tsx [hunk 1]
```diff
@@ -67,13 +82,14 @@ const Button: React.FC<Props & TestProps & EventProps> = props => {
         iconRight,
         childrenCount: React.Children.count(children),
       })}
-      onClick={onClick}
+      onClick={handleAsyncOperation}
```
Comment (kboutsika) [inline src/components/Button/Button.tsx:85]: I am so familiar with the codebase, so maybe i miss something. If there there is a `submit` button inside a form for example, so the form should be submitted with enter key or with submit event, right? If so, how we can set the button in the loading state?

File: src/components/Button/Button.tsx [hunk 2]
```diff
@@ -2,12 +2,20 @@
...
+  buttonSpanStyle,
+  buttonStyle,
+  centralizedLoader,
+  childrenWrapperStyle,
+  iconStyle,
+} from './Button.style';
 import { useTypeColorToColorMatch } from '../../hooks/useTypeColorToColorMatch';
+import { useLoading } from 'hooks/useLoading';
+import Loader from '../Loader';
+import { useRef } from 'react';
```
Comment (kboutsika) [inline src/components/Button/Button.tsx:18]: there is already everything imported from `react` in line 4
`import * as React from 'react';`. How about to do the below?
+ either `import * as React from 'react';` and `React. useRef<HTMLSpanElement>(null)`, or
+ `import React, { useRef } from 'react';`

File: src/components/Button/Button.tsx (hunk 1 above)
Comment (michalisKout) [reply src/components/Button/Button.tsx:85]: At the moment, our buttons do not have a specific type, which semantically incorrect. Although, based on the buttons' usage is not required for now.
I already planned to open a bug in the DS board. We can introduce a new prop buttonType ('submit', 'reset', 'button'). 

Then we can investigate accessing the loading state using keys or button types.

Comment (kboutsika) [reply src/components/Button/Button.tsx:85]: ok! thank you for the reply!

**Review by kboutsika — APPROVED** [review]

**Comment by panvourtsis** [comment]
reminder !

**Comment by michalisKout** [comment]
> reminder !

waiting for your approval on this :)

**Comment by Christos-Zacharopoulos** [comment]
> > reminder !
> 
> waiting for your approval on this :)

image

File: src/components/Loader/Loader.stories.mdx [hunk 3]
```diff
@@ -1,8 +1,9 @@
 import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
 
 import Loader from './Loader';
+import {loaderTypes} from './Loader.utils';
 
-<Meta title="Design System/Loader" component={Loader} />
+<Meta title="Design System/Loaders" component={Loader} />
```
Comment (panvourtsis) [inline src/components/Loader/Loader.stories.mdx:6]: why plural? In that case Button could be Buttons

File: src/components/Loader/components/Spinner/Spinner.tsx [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import React, { FC } from 'react';
+import { spinnerContainer } from './Spinner.style';
+import { generateTestDataId } from '../../../../utils/helpers';
+
+interface Props {
```
Comment (panvourtsis) [inline src/components/Loader/components/Spinner/Spinner.tsx:5]: extend TestId from utils. There is already a type for that

File: src/components/Loader/Loader.stories.mdx (hunk 3 above)
Comment (michalisKout) [reply src/components/Loader/Loader.stories.mdx:6]: I thought of this like a group of loaders. Like having different geometric shapes.

Geometric Shapes (plural) => Square, triangle, circle etc...
Loaders => Spinner, dots, indeterminate etc... 

but i can revert it to `Loader` if this is blocking

Comment (panvourtsis) [reply src/components/Loader/Loader.stories.mdx:6]: no its not blocking it was just confusing as it refers to a single point system like buttons

**Pull request merged**